icu4j/main/core/src/main/java/com/ibm/icu/impl/units/UnitPreferences.java

```diff
@@ -92,36 +89,39 @@ public UnitPreference[] getPreferencesFor(String category, String usage, ULocale
             }
         }
 
-        String region = locale.getCountry();
+        String region = ULocale.getRegionForSupplementalData(locale, false);
 
         // Check the locale system tag, e.g `ms=metric`.
         String localeSystem = locale.getKeywordValue("measure");
-        boolean isLocaleSystem = false;
-        if (measurementSystem.containsKey(localeSystem)) {
-            isLocaleSystem = true;
-            region = measurementSystem.get(localeSystem);
-        }
-
-        // Check the region tag, e.g. `rg=uszzz`.
-        if (!isLocaleSystem) {
-            String localeRegion = locale.getKeywordValue("rg");
-            if (localeRegion != null && localeRegion.length() >= 3) {
-                if (localeRegion.equals("default")) {
-                    region = localeRegion;
-                } else if (Character.isDigit(localeRegion.charAt(0))) {
-                    region = localeRegion.substring(0, 3); // e.g. 001
-                } else {
-                    // Capitalize the first two character of the region, e.g. ukzzzz or usca
-                    region = localeRegion.substring(0, 2).toUpperCase(Locale.ROOT);
-                }
-            }
-        }
+        boolean isLocaleSystem = measurementSystem.containsKey(localeSystem);
 
         String[] subUsages = getAllUsages(usage);
         UnitPreference[] result = null;
         for (String subUsage :
                 subUsages) {
             result = getUnitPreferences(category, subUsage, region);
+
+            if (result != null && isLocaleSystem) {
+                ConversionRates rates = new ConversionRates();
+                boolean unitsMatchSystem = true;
+                for (UnitPreference unitPref : result) {
+                    MeasureUnitImpl measureUnit = MeasureUnitImpl.forIdentifier(unitPref.getUnit());
+                    List<SingleUnitImpl> singleUnits = new ArrayList<>(measureUnit.getSingleUnits());
+                    for (SingleUnitImpl singleUnit : singleUnits) {
+                        String systems = rates.extractSystems(singleUnit);
+                        if (!systems.contains("metric_adjacent")) {
+                            if (!systems.contains(localeSystem)) {
+                                unitsMatchSystem = false;
+                            }
+                        }
```

Comment:
Shouldn't we allow `metric_adjacent` only when the `localeSystem` is `metric`? Because if we have a unit preference with `metric_adjacent` units but `imperial` was requested, then we should override with the imperial units, right?

Reply:
Well, the big problem I was having was with time units, which are all labeled `metric-adjacent`.  From looking at the other units, I didn't see any where `metric-adjacent` was being used in contradistinction to `ussystem` or `uksystem`, so this approach seemed safe.  An alternative approach might be to only look at the _first_ unit of a compound unit instead of all of them, or to explicitly avoid checking time units.  I didn't consider the first option, fearing I'd missing things that way, and the second option seemed too complicated (I couldn't see an easy way to tell whether something was a time unit).

Reply:
Okay, fair enough. I think this is code we can continue to tweak as more cases come up.